### DISCUSSION TITLE
Probable issue with object indices in `SelectionResult` replacing the `raw_indices` properly !!

### DISCUSSION THREAD

**Discussion by gsaha009** [columnflow/columnflow#582] [General]
Hello Experts,

First of all, I am sorry for asking this long question. 
I am trying to categorize my events in three jet multiplicity bins. Let me make it clear so that it can be followed quickly:

 - In my `Selection/physics_objects.py`, I am replacing the `jets` with `good_jet_indices` in the [SelectionResult](https://github.com/gsaha009/CPinHToTauTau/blob/main/httcp/selection/physics_objects.py#L525). I think this indices will replace the actual `Jet` collection after `Reduction`
 - Then at the end of the `jet_selection`, I am returning the [same indices](https://github.com/gsaha009/CPinHToTauTau/blob/main/httcp/selection/physics_objects.py#L537) too, because in my [`Selection/main.py`](https://github.com/gsaha009/CPinHToTauTau/blob/main/httcp/selection/main.py#L363), I want to build several event level masks.
 - in [build_abcd_masks](https://github.com/gsaha009/CPinHToTauTau/blob/main/httcp/selection/main.py#L369), I am building the `has_nJet` categories with the returned indices, as defined [here](https://github.com/gsaha009/CPinHToTauTau/blob/main/httcp/selection/event_category.py#L260-L262).
 - In the above, it is pretty straight forward, and I have checked that it is working
 - In my [`categorization/main`](https://github.com/gsaha009/CPinHToTauTau/blob/main/httcp/categorization/main.py#L44-L54), I have defined the jet categories using the event level masks
 - Now, in `production/main`, [here](https://github.com/gsaha009/CPinHToTauTau/blob/main/httcp/production/main.py#L108) I have defined the `n_jet` variable again using `events.Jet.pt` which I think should be the same as `good_jet_indices` that I saved in the [`SelectionResult`](https://github.com/gsaha009/CPinHToTauTau/blob/main/httcp/selection/physics_objects.py#L525) of `jet_selection`.
 - At the end, I am using the categories like [this](https://github.com/gsaha009/CPinHToTauTau/blob/main/httcp/config/categories.py#L158-L179)

BUT, the `njets` plots I am having for all `njet` category are totally weird. nJet category is showing less than n jets !!! 🤯 
For category : 
 - `tautau__real_1__hadD__has_0j` : [plot](https://gsahacmspoint.web.cern.ch/CPinHToTauTau/Plots/2022/PreFit/Run3_2022PreEE_full_20241130_190734/plot__proc_8_bce5a79095__cat_tautau__real_1__hadD__has_0j__var_n_jet.png)
 - `tautau__real_1__hadD__has_1j` : [plot](https://gsahacmspoint.web.cern.ch/CPinHToTauTau/Plots/2022/PreFit/Run3_2022PreEE_full_20241130_190734/plot__proc_8_bce5a79095__cat_tautau__real_1__hadD__has_1j__var_n_jet.png)
 - `tautau__real_1__hadD__has_2j` : [plot](https://gsahacmspoint.web.cern.ch/CPinHToTauTau/Plots/2022/PreFit/Run3_2022PreEE_full_20241130_190734/plot__proc_8_bce5a79095__cat_tautau__real_1__hadD__has_2j__var_n_jet.png)

What I think that the `events.Jet` collection is not replaced by the `good_jet_indices`. If you guys can have a look and give me any suggestions, it would be very helpful.
Thank you very much

Gourab

**Comment by mafrahm** [comment]
Hi @gsaha009, the SelectionResults are only applied during `cf.ReduceEvents`, meaning that when you create your categories (or run any other Producer), the plain Jet collection will be used. If this is your problem, there are two solutions:

- run the `category_ids` Producer on it's own or as part of some other Producer in `cf.ProduceColumns`
- you apply the object masks yourself, e.g. as done here with the `create_collections_from_masks` function as done in this example: https://gitlab.cern.ch/cms-analysis/analysisexamples/columnflow-demo/-/blob/master/h4l/production/example.py?ref_type=heads#L67-71

**Reply by gsaha009** [reply]
Hi @mafrahm 
I am creating the masks for my categories [here](https://github.com/gsaha009/CPinHToTauTau/blob/main/httcp/selection/main.py#L369-L373), and I am using the same [`good_jet_indices`](https://github.com/gsaha009/CPinHToTauTau/blob/main/httcp/selection/event_category.py#L259-L262) what I have included in the [`SelectionResults`](https://github.com/gsaha009/CPinHToTauTau/blob/main/httcp/selection/physics_objects.py#L525) already. 
So I am using the modified jet indices to build categories, and I am adding the same in the `SelectionResults`. Logically, this should work, no?
Thanks

**Reply by gsaha009** [reply]
Hi again...

please ignore my last comment. I can see in the documentation as well that this approach is okay. Can you please confirm one thing?
If I don't do [this](https://github.com/gsaha009/CPinHToTauTau/blob/main/httcp/selection/event_category.py#L102), as I wasn't not doing this before, is it possible that when I require `njet` in my categorization, it will take some arbitrary values?

Gourab

**Reply by mafrahm** [reply]
Hi @gsaha, adding columns to `produces` just enables you to store the column on disk after the task is completed. If you use the column directly, it should not matter if you add it to the produces or not. In any case, if you try to use a column that is not existent, you should just get an error.

Regarding your issues, I am not able to see what you might be doing wrong from you code since this is quite a lot of code. Here are some ideas how you could try debugging this:

- Have you looked into your jet collection after reduction and ensured that you really only keep the jets you are interested in?
- You could also directly look into the events of some specific categories with this function: https://github.com/columnflow/columnflow/blob/e12b75bf66864b7d478fc26e8259c5940b6cede7/columnflow/config_util.py#L35
- You could also try calculating your category_ids after selection (in cf.ProduceColumns) to get rid of inconsistencies in some object definitions. I'd personally prefer this anyways since this allows you to redefine your categories without having to run your event selection again

**Reply by gsaha009** [reply]
Hi @mafrahm 

 - Have you looked into your jet collection after reduction and ensured that you really only keep the jets you are interested in?
   - Ans: Yes. The good jets replaced the raw jet collection and that was exactly what I needed
 - You could also directly look into the events of some specific categories with this function
   - yes, I have looked at it. I extracted events for a specific category, and I have found a lot of events are there from other categories. 
 - Calculating your category_ids after selection (in cf.ProduceColumns) 
   - Done it. Thanks for this suggestion. It did not solve the problem, but it made the code optimized.
   
I have figured out one thing though
This problem is coming when I am combining the njets and dm categories like [here](https://github.com/gsaha009/CPinHToTauTau/blob/main/httcp/config/categories.py#L167-L177). If I keep `dm` only, and remove `njets`, the decay mode plots look perfect, and if I remove `dm` and keep `njets`, the njets plots look good. So, the problem is coming from the combination of the twos. But, all are orthogonal to each other, as defined [here](https://github.com/gsaha009/CPinHToTauTau/blob/main/httcp/selection/event_category.py#L215-L255)
Although, if I define any of the root to leaf full category explicitly by hand, it works fine. Is there any ambiguity in the `create_category_combination` function?

**Comment by gsaha009** [comment]
Hi @mafrahm 

I have found the issue. I chose `category_ids` quite close ([L1](https://github.com/gsaha009/CPinHToTauTau/blob/main/httcp/config/categories.py#L62-L65), [L2](https://github.com/gsaha009/CPinHToTauTau/blob/main/httcp/config/categories.py#L105-L109)) to each other, and that was why it was skipping categories or merging other categories while combining, that I am not very sure about. I thought there would be some warning, but there was none. A warning in the `create_category_combination` would be helpful. I chose the ids carefully now, and the problem is gone. Thanks for your help.

Gourab

**Reply by mafrahm** [reply]
Hi @gsaha009, thank you for figuring this out! In theory, category IDs should be unique but it seems that this check does not work with nested categories. While this might be mostly an issue with order, I think it should be easy to add a simple uniqueness check to the `create_category_combinations` function.

**Reply by gsaha009** [reply]
Hi @mafrahm 
yes, I agree. This is related to `order`. I will add a check for uniqueness in my category definition.
